allennlp/training/trainer.py

```diff
@@ -399,7 +399,20 @@ def _data_parallel(self, batch):
         of torch.nn.parallel.data_parallel to support the allennlp model
         interface.
         """
+        metadata_batch_size = len(batch['metadata']) if 'metadata' in batch and isinstance(batch['metadata'],list) else None
```

Comment:
I'd really like to decouple the multiple GPU functionality from any of our Iterator specific code to keep the code as modular as possible.  I've been happily training on multiple GPUs using `Trainer` without using the allennlp iterators.

Reply:
Is it sufficient to require any usage of `MetadataField` and multiple GPU training to use the `metadata` key? Or perhaps more generally, add an option to `Trainer` that specifies the additional metadata keys to also scatter (that defaults to just the `metadata` key)?

Reply:
Except, with @joelgrus working on getting multi-processing into the data iterator, that's where tensors are constructed on the GPU - the multi-processing is pretty pointless if you then have to move the constructed tensors onto another device.  That code really should be aware of which GPU to put things on for data parallel stuff, and it seems pretty straightforward to do it in a smart way.

But even if we don't do it that way, you really don't want to have to pass in any additional keys to the `Trainer`, or have arbitrary restrictions on the keys that you can use.  The code that I linked to in pytorch really isn't very complicated; we have very similar blocks in several places in our code.  We'd just have to implement the `scatter` to different GPUs ourselves.

Reply:
I was talking about this with @vidurj on slack, but slack appears to be down for me right now, so continuing here.  Basically, you need a top-level type-check like in the pytorch method.  If the item is a tensor, you do the scatter.  If it's a dictionary, you do the scatter like the pytorch method does.  If it's a _list of length `batch_size`_, you need to do something different: instead of scattering all of the things inside the list, you scatter the list itself, then recurse on each list _with the device that you picked_, so tensors inside the list get moved to the right device.

Reply:
That makes sense. Do we need to worry about a list happening to have length `batch_size` but not actually representing a batch?

Reply:
I don't think so.  Because of the way the `Batch` is constructed, that should never happen.

Reply:
Also, to @matt-peters's earlier point, if we don't want to be creating lots of tensors on the GPU during multi-processed batch creation, we need to change how out data iterator works.  This is very much related to what @joelgrus said he was working on right now, and needs some thought.

Reply:
jumping in kind of late here, I just pushed up a not-very-full-featured `MultiprocessIterator`. I couldn't figure out a clean way to add the functionality to Basic or Bucket, so I just made it its own class, it doesn't have "Bucket" features yet but it could:

https://github.com/allenai/allennlp/pull/1723/files#diff-7534c31c53ce78f753a0f71c30007a0d